### DISCUSSION TITLE
Prevent server-side network call with createTRPCProxyClient

### DISCUSSION THREAD

**Discussion by austins** [trpc/trpc#2735] [Q&A]
I'm working on integrating the tRPC v10 fetch adapter with SvelteKit.

There's multiple ways you can fetch initial data in SvelteKit:
1. `+page.server.ts`: server-side only (here we can use `router.createCaller` to avoid a network call).
2. `+page.svelte`: client-side only (browser's fetch).
3. `+page.ts`: hybrid of server-side (using [SvelteKit's fetch implementation](https://kit.svelte.dev/docs/load#input-methods-fetch) from the load function) and client-side (browser's fetch).

The third way gives you the best of both worlds: SSR for initial data and client-side when navigated to client-side to avoid render blocking (similar to tRPC's Next.js SSR implementation with `react-ssr-prepass` and `getInitialProps`).

Here is how I'm using createTRPCProxyClient: [+page.ts](https://github.com/austins/trpc-sveltekit-fetchadapter-example/blob/3a171630f102b87923e09736720dcdf8d4a2ece3/src/routes/hybrid/%2Bpage.ts#L8) with [createTRPCProxyClient](https://github.com/austins/trpc-sveltekit-fetchadapter-example/blob/f9908f759473f32efbcbdaea4e72bc76a2f4b9a6/src/lib/trpc/trpcClient.ts#L8).

The result is that when you first load the page, it **makes a network call server-side** then serializes the result to the view and when you navigate to the page client-side, it makes a client-side network call.

We can use `import { browser } from "$app/environment";` to detect if the request is server-side or client-side, but for the next step, I'm interested to know if anyone has ideas for how to avoid a network call server-side.

Ideas:
- Can we override the fetch on a server-side request and make it run createCaller (although the signature is different since it doesn't have `.query()`, `.mutation()`, and so on like the proxy client)? Although, `+page.ts` doesn't provide the `request` object, SvelteKit is able to make its own fetch implementation have "internal requests (e.g. for +server.js routes) go direct to the handler function when running on the server, without the overhead of an HTTP call",
- Use both `+page.server.ts` and `+page.ts` (receives the data prop from `+page.server.ts`), although this can cause somewhat duplicate code and isn't the most convenient DX.

**Comment by austins** [comment]
Actually, `+page.server.ts` seems to not cause a full page load and only runs the logic within it when navigating to a page that uses only that. However, technically, this is still render blocking since the whole page needs the data from it before being able to render any components within it.

**Comment by KATT** [comment]
This sounds like reaaally nice DX and UX :)

Here's an idea to do something universal: In the `+page.ts` you could write an ending [link](https://trpc.io/docs/v10/links#creating-a-custom-link) that doesn't go through HTTP boundaries. 

What you would want to end up is probably something like this:

```ts
import type { AppRouter } from '@/server/routers/app';
// <>make sure this doesn't get included for the browser
import { appRouter } from '@/server/routers/app';
import { ssrLink } from 'trpc-sveletekit';
// </>
import { httpBatchLink } from '@trpc/client';
import { createTRPCNext } from '@trpc/next';

export const trpc = createTRPCNext<AppRouter>({
  config() {
    return {
      links: [
        typeof window === 'undefined'
          ? ssrLink({
              router: appRouter,
            })
          : httpBatchLink({
              url: 'http://localhost:3000/api/trpc',
              maxURLLength: 2083, // a suitable size
            }),
      ],
    };
  },
});

```
